# Business Understanding

Problem Statement :
Skylard Ltd. is interested in purchasing and operating airplanes for commercial and private enterprises. The aim is to analyze aviation accidents data to provide insights and recommendations on which aircrafts are the lowest risk for the company.

Objectives:
1.Identify trends and patterns in aviation accident data to determine which aircraft models have the highest vs lowest accident and incident rates and highest vs lowest level of damage to the aircraft.
2.Evaluate injury severity by aircraft models to assess overall risk levels per model.
3.Develop  recommendations on the safest aircraft models for fleet acquisition.

# Data Understanding

Dataset Source: National Transportation Safety Board (1962 – 2023);
Size: 97,404 rows, 31 columns;
Focus: Civil aviation accidents and incidents (US and international waters)


# Data Cleaning

In [18]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

In [13]:
df = pd.read_csv("AviationData.csv", encoding='latin1')

In [14]:
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                    2   
1          NaN  ...          Personal         NaN                    4   
2          NaN  ...          Personal         NaN                    3   
3          NaN  ...          Personal         NaN                    2   
4          NaN  ...          Personal         NaN                    1   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                      0                    0               0   
1                      0                    0               0   
2                    NaN                  NaN             NaN   
3                      0                    0               0   
4                      2                  NaN               0   

  Weather.Condition Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                Cruise  Probable Cause              NaN  
1               UNK               Unknown  Probable Cause       19-09-1996  
2               IMC                Cruise  Probable Cause       26-02-2007  
3               IMC                Cruise  Probable Cause       12-09-2000  
4               VMC              Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [15]:
df.tail()

Event.Id Investigation.Type Accident.Number  Event.Date  \
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
88884  Annapolis, MD  United States      NaN       NaN          NaN   
88885    Hampton, NH  United States      NaN       NaN          NaN   
88886     Payson, AZ  United States  341525N  1112021W          PAN   
88887     Morgan, UT  United States      NaN       NaN          NaN   
88888     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
88884                    0                      1                    0   
88885                    0                      0                    0   
88886                    0                      0                    0   
88887                    0                      0                    0   
88888                    0                      1                    0   

      Total.Uninjured Weather.Condition Broad.phase.of.flight Report.Status  \
88884               0               NaN                   NaN           NaN   
88885               0               NaN                   NaN           NaN   
88886               1               VMC                   NaN           NaN   
88887               0               NaN                   NaN           NaN   
88888               1               NaN                   NaN           NaN   

      Publication.Date  
88884       29-12-2022  
88885              NaN  
88886       27-12-2022  
88887              NaN  
88888       30-12-2022  

[5 rows x 31 columns]

In [17]:
df.shape

(88889, 31)

 #Observation: Power BI reports 97,404 rows in the original CSV file while pandas loaded only 88,889 rows. 
 #Observation: This suggests around 8,515 rows were skipped due to malformed rows or corrupt data.

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Event.Id                88889 non-null  object
 1   Investigation.Type      88889 non-null  object
 2   Accident.Number         88889 non-null  object
 3   Event.Date              88889 non-null  object
 4   Location                88837 non-null  object
 5   Country                 88663 non-null  object
 6   Latitude                34382 non-null  object
 7   Longitude               34373 non-null  object
 8   Airport.Code            50249 non-null  object
 9   Airport.Name            52790 non-null  object
 10  Injury.Severity         87889 non-null  object
 11  Aircraft.damage         85695 non-null  object
 12  Aircraft.Category       32287 non-null  object
 13  Registration.Number     87572 non-null  object
 14  Make                    88826 non-null  object
 15  Mo

In [22]:
df.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Columns: 31 entries, Event.Id to Publication.Date
dtypes: object(31)
memory usage: 21.0+ MB


In [23]:
df.describe()

Event.Id Investigation.Type Accident.Number  Event.Date  \
count            88889              88889           88889       88889   
unique           87951                  2           88863       14782   
top     20001212X19172           Accident      WPR22LA143  1984-06-30   
freq                 3              85015               2          25   

             Location        Country Latitude Longitude Airport.Code  \
count           88837          88663    34382     34373        50249   
unique          27758            219    25589     27154        10375   
top     ANCHORAGE, AK  United States  332739N  0112457W         NONE   
freq              434          82248       19        24         1488   

       Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
count         52790  ...             82697       16648                77488   
unique        24871  ...                26       13590                  125   
top         Private  ...          Personal       Pilot                    0   
freq            240  ...             49448         258                59675   

       Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
count                   76379                76956           82977   
unique                     50                   57             379   
top                         0                    0               0   
freq                    63289                61454           29879   

       Weather.Condition Broad.phase.of.flight   Report.Status  \
count              84397                 61724           82508   
unique                 4                    12           17075   
top                  VMC               Landing  Probable Cause   
freq               77303                 15428           61754   

       Publication.Date  
count             75118  
unique             2924  
top          25-09-2020  
freq              17019  

[4 rows x 31 columns]

In [28]:
df.describe().T

count unique             top   freq
Event.Id                88889  87951  20001212X19172      3
Investigation.Type      88889      2        Accident  85015
Accident.Number         88889  88863      WPR22LA143      2
Event.Date              88889  14782      1984-06-30     25
Location                88837  27758   ANCHORAGE, AK    434
Country                 88663    219   United States  82248
Latitude                34382  25589         332739N     19
Longitude               34373  27154        0112457W     24
Airport.Code            50249  10375            NONE   1488
Airport.Name            52790  24871         Private    240
Injury.Severity         87889    109       Non-Fatal  67357
Aircraft.damage         85695      4     Substantial  64148
Aircraft.Category       32287     15        Airplane  27617
Registration.Number     87572  79105            NONE    344
Make                    88826   8237          Cessna  22227
Model                   88797  12318             152   2367
Amateur.Built           88787      2              No  80312
Number.of.Engines       82805      7               1  69582
Engine.Type             81812     13   Reciprocating  69530
FAR.Description         32023     31             091  18221
Schedule                12582      3            NSCH   4474
Purpose.of.flight       82697     26        Personal  49448
Air.carrier             16648  13590           Pilot    258
Total.Fatal.Injuries    77488    125               0  59675
Total.Serious.Injuries  76379     50               0  63289
Total.Minor.Injuries    76956     57               0  61454
Total.Uninjured         82977    379               0  29879
Weather.Condition       84397      4             VMC  77303
Broad.phase.of.flight   61724     12         Landing  15428
Report.Status           82508  17075  Probable Cause  61754
Publication.Date        75118   2924      25-09-2020  17019

#Observation: The highest investigation type is accident in this dataset is accident - an accident in aviation involves substantial aircraft damage.
#Observation: The most common aircraft make and model in this dataset Cessna 152.

In [32]:
df1 = df.copy(deep=True)

In [33]:
df1.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                    2   
1          NaN  ...          Personal         NaN                    4   
2          NaN  ...          Personal         NaN                    3   
3          NaN  ...          Personal         NaN                    2   
4          NaN  ...          Personal         NaN                    1   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                      0                    0               0   
1                      0                    0               0   
2                    NaN                  NaN             NaN   
3                      0                    0               0   
4                      2                  NaN               0   

  Weather.Condition Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                Cruise  Probable Cause              NaN  
1               UNK               Unknown  Probable Cause       19-09-1996  
2               IMC                Cruise  Probable Cause       26-02-2007  
3               IMC                Cruise  Probable Cause       12-09-2000  
4               VMC              Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

# Dropping all unnecessary columns 

In [42]:
df1.drop(['Latitude', 'Longitude','Airport.Code','Airport.Name', 'Schedule', 'Accident.Number', 'Registration.Number', 'FAR.Description', 'Report.Status'], axis = 1, inplace=True)

KeyError: "['Latitude' 'Longitude' 'Airport.Code' 'Airport.Name' 'Schedule'\n 'Accident.Number' 'Registration.Number' 'FAR.Description' 'Report.Status'] not found in axis"

In [44]:
df1.shape

(88889, 22)

In [45]:
df1.columns

Index(['Event.Id', 'Investigation.Type', 'Event.Date', 'Location', 'Country',
       'Injury.Severity', 'Aircraft.damage', 'Aircraft.Category', 'Make',
       'Model', 'Amateur.Built', 'Number.of.Engines', 'Engine.Type',
       'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Publication.Date'],
      dtype='object')

#Drop additional columns and remain with those most relevant to the analysis.

In [47]:
df1.drop(['Location', 'Air.carrier','Injury.Severity','Publication.Date'], axis = 1, inplace=True)

# Filling in null values

In [48]:
df1.isnull().sum()

Event.Id                      0
Investigation.Type            0
Event.Date                    0
Country                     226
Aircraft.damage            3194
Aircraft.Category         56602
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7077
Purpose.of.flight          6192
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
dtype: int64

# Filling in categorical variables using mode 

In [69]:
country_mode =df1.Country.mode()
country_mode

0    United States
dtype: object

In [59]:
df1["Country"].fillna(country_mode[0], inplace=True)

In [62]:
aircraftdamage_mode =df1["Aircraft.damage"].mode()
aircraftdamage_mode

0    Substantial
dtype: object

In [63]:
df1["Aircraft.damage"].fillna(aircraftdamage_mode[0], inplace=True)

In [66]:
aircraftcategory_mode =df1["Aircraft.Category"].mode()
aircraftcategory_mode

0    Airplane
dtype: object

In [67]:
df1["Aircraft.Category"].fillna(aircraftcategory_mode[0], inplace=True)

In [70]:
make_mode =df1.Make.mode()
make_mode

0    Cessna
dtype: object

In [71]:
df1["Make"].fillna(make_mode[0], inplace=True)

In [74]:
model_mode =df1.Model.mode()
model_mode

0    152
dtype: object

In [75]:
df1["Model"].fillna(model_mode[0], inplace=True)

In [77]:
amateurbuilt_mode =df1["Amateur.Built"].mode()
amateurbuilt_mode

0    No
dtype: object

In [78]:
df1["Amateur.Built"].fillna(amateurbuilt_mode[0], inplace=True)

In [80]:
enginetype_mode =df1["Engine.Type"].mode()
enginetype_mode

0    Reciprocating
dtype: object

In [81]:
df1["Engine.Type"].fillna(enginetype_mode[0], inplace=True)

In [83]:
purposeofflight_mode =df1["Purpose.of.flight"].mode()
purposeofflight_mode

0    Personal
dtype: object

In [84]:
df1["Purpose.of.flight"].fillna(purposeofflight_mode[0], inplace=True)

In [86]:
weathercondition_mode =df1["Weather.Condition"].mode()
weathercondition_mode

0    VMC
dtype: object

In [87]:
df1["Weather.Condition"].fillna(weathercondition_mode[0], inplace=True)

In [88]:
broadphaseofflight_mode =df1["Broad.phase.of.flight"].mode()
broadphaseofflight_mode

0    Landing
dtype: object

In [89]:
df1["Broad.phase.of.flight"].fillna(broadphaseofflight_mode[0], inplace=True)

In [90]:
df1.isnull().sum()

Event.Id                      0
Investigation.Type            0
Event.Date                    0
Country                       0
Aircraft.damage               0
Aircraft.Category             0
Make                          0
Model                         0
Amateur.Built                 0
Number.of.Engines          6084
Engine.Type                   0
Purpose.of.flight             0
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition             0
Broad.phase.of.flight         0
dtype: int64

# Filling in numerical variables using mean/median

In [104]:
df1.dtypes

Event.Id                  object
Investigation.Type        object
Event.Date                object
Country                   object
Aircraft.damage           object
Aircraft.Category         object
Make                      object
Model                     object
Amateur.Built             object
Number.of.Engines         object
Engine.Type               object
Purpose.of.flight         object
Total.Fatal.Injuries      object
Total.Serious.Injuries    object
Total.Minor.Injuries      object
Total.Uninjured           object
Weather.Condition         object
Broad.phase.of.flight     object
dtype: object

#Observation: The numeric columns have been read as strings. 
#Change 5 columns to numeric - Number.of.Engines, Total.Fatal.Injuries, Total.Serious.Injuries, Total.Minor.Injuries, Total.Uninjured

In [105]:
cols_to_convert = ['Number.of.Engines', 'Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured']

In [106]:
for col in cols_to_convert:
    df1[col] = pd.to_numeric(df1[col], errors='coerce')

In [108]:
df1[cols_to_convert].dtypes

Number.of.Engines         float64
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
dtype: object

In [109]:
df1.dtypes

Event.Id                   object
Investigation.Type         object
Event.Date                 object
Country                    object
Aircraft.damage            object
Aircraft.Category          object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
Purpose.of.flight          object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
dtype: object